### ISSUE TITLE
[Bug] Missing data points when combine facet_wrap() with both the color and shape aesthetics

### ISSUE THREAD

**Issue by 18874studentvgu** [Vindaar/ggplotnim#180]
# How to replicate:
Not sure if the alpine image is the problem or not so I'm listing it here, try adding these steps if you can't reproduce the bug
* [alpine docker image of nim](https://hub.docker.com/layers/nimlang/nim/2.0.4-alpine/images/sha256-0f267eb6014d34bfc8bb9cee68fb5e2ab1fb27ec28c47a432d8b55b1d9086602?context=explore)
* Enter shell: `docker run --rm -it -v .:/usr/src/app -w /usr/src/app nim:2.0.4-alpine /bin/sh -c "sh; echo '[Exiting]'; exit 0"`
* `nimble refresh`, install datamancer and ggplotnim
* Install cairo and other packages required by ggplotnim with apk if the compiler yells at you

## Data:
[Studentenstatistik_BB_Datensatz.csv](https://github.com/user-attachments/files/15804436/Studentenstatistik_BB_Datensatz.csv) (Modified from [here](https://data.europa.eu/data/datasets/7e0d0367-d0e9-45f7-8414-671bb84d0d7c?locale=en))

## The problematic code:
```nim
import datamancer
import ggplotnim

let df = readCsv("csv/Studentenstatistik_BB_Datensatz.csv",sep=';')
ggplot(df, aes(y="Counting",x="Winter_semester",color ="Subject_group",shape="Type_of_university"))  + 
    geom_point() +
    facet_wrap(["Gender"]) + 
    ggsave("graph/buggy2.svg",width = 1080.0,height=720)
```
## Output:
![buggy2](https://github.com/Vindaar/ggplotnim/assets/133742827/e86d96bd-c655-4fff-ba4c-b54c7a3ec784)

Compared to the equivalent code using `plotnine`, an entire green to blue color range is gone!
![ref2](https://github.com/Vindaar/ggplotnim/assets/133742827/1418d62c-038c-4b70-b54a-2cfb1c8efb8c)

## Additional information:
I did try to reproduce the bug with some combination of the 3 elements, to no avail. The bug only seems to appear when both the shape and color aesthetics alongside the facet_wrap command is present in the plotting chain.
* No facet:
    ![buggy0](https://github.com/Vindaar/ggplotnim/assets/133742827/de9e5bf5-533f-45ee-aba4-856d1a39a23d)
* Either `shape` or `color` is missing:
    ![buggy6](https://github.com/Vindaar/ggplotnim/assets/133742827/b9d59762-59fe-4dfe-aa12-113710a699ad)
    ![buggy5](https://github.com/Vindaar/ggplotnim/assets/133742827/9b09c3cc-bfd7-4150-b6b2-ede5033fc2c2)

I feel like this could need a bit of an investigation.

**Comment by Vindaar** [comment]
Hey!

Thanks for the detailed report. Huh, that is indeed funny. I'll try to look into it soon!

**Comment by Vindaar** [comment]
Ouch, I think it's (at least partially) a nasty bug in Datamancer's `arrange` for more than 2 columns. Investigating...

**Comment by Vindaar** [comment]
The issue was indeed "just" the `arrange` bug in Datamancer. Unfortunately right now I'm fighting an ORC / ARC bug, which causes a segfault on (it seems) everything but devel in the new logic... :/

**Comment by Vindaar** [comment]
I'm unsure about what to do right now. I _hope_ that the fix to the regression that is there in Nim 2.0 was already backported and will show up in a new release soon. Fortunately, at least the release of 2.2 should also be very close.

I can merge the Datamancer PR linked above, but given that you are on 2.0.4 it won't actually fix your problem (it'll just make it work by breaking every `arrange` call). Hmm. If I come up with something better, I'll comment here.

**Comment by Vindaar** [comment]
With Nim 2.0.6 released, I can finally merge the Datamancer PR. Once that is done, please update your datamancer version and to Nim 2.0.6 and let me know if the plot looks as expected (it does here).

**Comment by Vindaar** [comment]
Datamancer version `v0.4.5` is tagged now. Please let me know if it all works after updating!

**Comment by Vindaar** [comment]
Here's your plot with Datamancer `v0.4.5` and some minor "prettification":

```nim
import ggplotnim

const path = "~/CastData/ExternCode/Datamancer/data/Studentenstatistik_BB_Datensatz.csv"

let df = readCsv(path,sep=';')
ggplot(df, aes("Winter_semester", "Counting", color = "Subject_group", shape = "Type_of_university")) +
  geom_point() +
  facet_wrap(["Gender"]) +
  margin(right = 15) +
  scale_x_continuous(breaks = @[2010, 2015]) +
  xlab("Winter semester") +
  discreteLegendWidth(0.5) + discreteLegendHeight(0.5) +
  legendFont(tickFont = font(8.0)) +
  ggsave("/tmp/fixed.pdf", width = 1080, height = 480)
```

(converted to a PNG for GH):

![fixed](https://github.com/Vindaar/ggplotnim/assets/7742232/0ac66a2b-5058-4b48-ac75-0776db269db0)


I'm closing the issue for now. Feel free to reopen / open a new issue if you need any help!

### Just a minor hint:

If you wish to adjust the way a plot looks, let me point you to the `Theme` type:
https://github.com/Vindaar/ggplotnim/blob/master/src/ggplotnim/ggplot_types.nim#L395-L485

You can either adjust it in the way I've done up there by calling individual functions, or you can even just define a TOML file that contains the Theme adjustments. This example here is from recently when I had to produce some plots for a poster (hence the huge fonts and weird colors):

```toml
[Theme]
titleFont = "font(18.0)"
labelFont = "font(18.0)"
tickLabelFont = "font(14.0)"
tickLength = 10.0
tickWidth = 2.0
gridLineWidth = 2.0
legendFont = "font(14.0)"
legendTitleFont = "font(18.0, bold = true)"
facetHeaderFont = "font(18.0, alignKind = taCenter)"
baseLabelMargin = 0.5
annotationFont = "font(9.0, family = \"monospace\")"
continuousLegendHeight = 2.2
continuousLegendWidth = 0.5
discreteLegendHeight = 0.6
discreteLegendWidth = 0.6
plotMarginRight = 5.0
plotMarginLeft = 3.0
plotMarginTop = 1.0
plotMarginBottom = 2.5
canvasColor = "#7fa7ce"
baseScale = 1.5
``` 

which you then use in your code by passing a:

```nim
tomlTheme("path_to_file.toml")
```
